### PR TITLE
INT-7445 Fix duplicate key on applications and groups relationship

### PR DESCRIPTION
**Pull request by jroblesx**
### Description
An error is generated while adding a group and application relationship, to mitigate the error I used `jobState.hasKey` to verify if exists before try to add an existing relationship.

### CODE REVIEW

File: src/steps/applications.ts [hunk 1]
```diff
@@ -48,16 +48,19 @@ export async function fetchApplications({
...
-          createApplicationGroupRelationships(
-            appEntity,
-            group,
-            createOnInvalidRoleFormatFunction(logger, {
-              appId,
-              groupId: group.id,
-            }),
-          ),
-        );
+        const relationshipId = `${group.id}|assigned|${appEntity._key}`;
```
Comment (mishelashala) [inline src/steps/applications.ts:51]: can we add a unit test and an abstraction for this?

```ts
const relationShipId = getRelationShipId(group.id, appEntity._key);
```

**Review by mishelashala — CHANGES_REQUESTED** [review]

File: src/steps/applications.ts (hunk 1 above)
Comment (ndowmon) [reply src/steps/applications.ts]: TBH, this doesn't seem like complex enough logic that it's worth unit testing/abstracting, in my opinion.

File: src/steps/applications.ts [hunk 2]
```diff
@@ -48,16 +49,24 @@ export async function fetchApplications({
...
-            group,
-            createOnInvalidRoleFormatFunction(logger, {
-              appId,
-              groupId: group.id,
-            }),
-          ),
+        const relationshipId = buildApplicationGroupRelationshipId(
+          group.id,
+          appEntity._key,
+          logger,
```
Comment (ndowmon) [inline src/steps/applications.ts:55]:
```suggestion
      const relationship = createApplicationGroupRelationships(
            appEntity,
            group,
            createOnInvalidRoleFormatFunction(logger, {
              appId,
              groupId: group.id,
            }),
          )
```

File: src/steps/applications.ts [hunk 3]
```diff
@@ -48,16 +49,24 @@ export async function fetchApplications({
...
+              appEntity,
+              group,
+              createOnInvalidRoleFormatFunction(logger, {
+                appId,
+                groupId: group.id,
+              }),
+              logger,
+            ),
+          );
+        }
```
Comment (ndowmon) [inline src/steps/applications.ts:69]:
```suggestion
        for (const relationship of relationships) {
          if (!jobState.hasKey(relationship._key)) {
            await jobState.addRelationships(relationship)
          }

        }
```

Comment (ndowmon) [reply src/steps/applications.ts:69]: I would also consider adding a comment about _why_ multiple assignments are possible. 

E.g.

```ts
        /**
         * Multiple relationships for the same group can be encountered if the 
         * app has specific profiles associated. 
         * 
         * For example, the AWS Account Federation app has a profile that 
         * specifies "role". The same group could be assigned multiple roles, 
         * such as `Developer` and `DeveloperExternal`.
         * 
         *  {
         *    id: '00gc3frhdodo90K6Q4x7',
         *    lastUpdated: '2023-04-07T19:19:30.000Z',
         *    priority: 0,
         *    profile: { role: 'Developer' },
         *  }
         *  {
         *    id: '00gc3frhdodo90K6Q4x7',
         *    lastUpdated: '2023-04-07T19:19:30.000Z',
         *    priority: 1,
         *    profile: { role: 'DeveloperExternal' },
         *  }
         */
```

Comment (jroblesx) [reply src/steps/applications.ts]: Added!

File: src/steps/applications.ts [hunk 4]
```diff
@@ -48,16 +48,26 @@
...
+            appId,
+            groupId: group.id,
+          }),
         );
+
+        /**
+         * Multiple relationships for the same group can be encountered if the
+         * app has specific profiles associated.
+         */
+        if (relationships && relationships.length) {
```
[the remark this answers is not in the document]
Comment (jroblesx) [reply src/steps/applications.ts]: If clause removed

**Review by ndowmon — APPROVED** [review]

**Pull request closed without merging**